fladle-plugin/src/main/java/com/osacky/flank/gradle/FulladlePlugin.kt

```diff
@@ -138,3 +45,127 @@ class FulladlePlugin : Plugin<Project> {
+
+          outputs.configureEach {
+            flankGradleExtension.additionalTestApks.add(
+              project.rootProject.provider {
+                listOf("- test: $outputFile", debugApk, maxTestShards, clientDetails, environmentVariables)
+                  .filter { it.isNotEmpty() }
+                  .joinToString("\n")
+                  .trimStart()
+                  .trimEnd()
+              }
+            )
+          }
+        } else {
+          println("Ignoring $name test variant in $path: No tests in $projectDir/src/androidTest")
+        }
+      }
+    }
+  }
+}
+
+fun configureApplicationModule(project: Project, flankGradleExtension: FlankGradleExtension) = project.run {
+  pluginManager.withPlugin("com.android.application") {
+    val fulladleModuleExtension = extensions.getByType(FulladleModuleExtension::class.java)
+    if (!fulladleModuleExtension.enabled.get())
+      return@withPlugin
+    val appExtension = extensions.getByType<AppExtension>()
+    // Only configure the first test variant per module.
+    // Does anyone test more than one variant per module?
+    var addedTestsForModule = false
+
+    // TODO deal with ignored/filtered variants
+    appExtension.testVariants.configureEach testVariant@{
+      if (addedTestsForModule) {
+        return@testVariant
+      }
+      val appVariant = testedVariant
+      appVariant.outputs.configureEach app@{
+
+        this@testVariant.outputs.configureEach test@{
+          val strs = mutableListOf<String>()
+          // TODO is this racy?
+          // If the debugApk isn't yet set, let's use this one.
+          if (!flankGradleExtension.debugApk.isPresent) {
+            flankGradleExtension.debugApk.set(rootProject.provider { this@app.outputFile.absolutePath })
+          } else {
+            // Otherwise, let's just add it to the list.
+            strs.add("- app: ${this@app.outputFile}")
+          }
+
+          // If the instrumentation apk isn't yet set, let's use this one.
+          if (!flankGradleExtension.instrumentationApk.isPresent) {
+            flankGradleExtension.instrumentationApk.set(rootProject.provider { this@test.outputFile.absolutePath })
+          } else {
+            // Otherwise, let's just add it to the list.
+            strs.add("      test: ${this@test.outputFile}")
+          }
+          val maxTestShards = fulladleModuleExtension.maxTestShards.let {
+            buildString {
+              if (it.isPresent) { append("    ") }
+              appendProperty(it, name = "max-test-shards")
+            }
+          }.trimEnd()
+
+          if (strs.isEmpty() || strs[0].isEmpty()) // root module, should not be added as additional test apk
+            return@test
+
+          val clientDetails = mapPropertyToYaml(fulladleModuleExtension.clientDetails, "client-details")
+
+          val environmentVariables = mapPropertyToYaml(fulladleModuleExtension.environmentVariables, "environment-variables")
```

Comment:
there's a bit of duplicated/copied logic between the two cases, can we extract the common configuration to a single method?

Reply:
Yeah, `mapPropertyToYaml` was an attempt to to extract out duplicate code, since there were two properties using it. I understand more work can be done in deduplication, but there might be diminishing returns.

Reply:
I extracted out a bit more of the duplicated logic into methods, have a look!